### DISCUSSION TITLE
Make Timer.one_shot==true by default

### DISCUSSION THREAD

**Discussion by PixelTim** [godotengine/godot-proposals#10821] [General Discussions]
There are different situations that make you turn this variable turn on or off, but I think that timer is mostly used without looping (`one_shot==true`), the examples are: player temporary invurnability after hit, temporary buff/debuff from a potion, cooldown of a skill, etc., so I think that it would be good to make `one_shot` proberty `true` by default. Of course, there still are situations where timer needs to be looped (`one_shot==false`), examples: day/night cycle, periodically spawning something, etc., but in my case, I set `one_shot==true` to most of my timers.

**Comment by AThousandShips** [comment]
I'd say most users don't use timers this way most of the time, most of the time for one-shot purposes people would use `get_tree().create_timer`

But regardless this can't be done because it breaks compatibility and no one could use `Timer` correctly if their project had `one_shot == false` as it would reset to `one_shot == true`

**Reply by PixelTim** [reply]
> But regardless this can't be done because it breaks compatibility and no one could use Timer correctly if their project had one_shot == false as it would reset to one_shot == true

Here's my idea to not break the compability: we will need a converter. If scene file has no line that sets `one_shot` property, this means that the node uses default value (i.e. false), and the converter will add `one_shot = false` line to the scene file, preventing this bug.

**Reply by AThousandShips** [reply]
That won't work, it can't know this, how would it tell if the file was before this change or not?

**Reply by PixelTim** [reply]
There is no need to tell the converter if the file was before this change or not, because it will modify **only** files where line `one_shot = ...` **doesn't exist**. In process of modifying it adds `one_shot = false` line, so if it will try to modify the same scene file again, it won't modify it because all timers will **already** have lines `one_shot = ...`, because the converter added those lines before. Also, the converter will work only once, when upgrating project, that had `one_shot = false` by default, to godot version which has `one_shot = true` by default (e.g. from 4.3 to 4.4).

**Reply by Zireael07** [reply]
You are assuming that one_shot line not existing means it's the default?

**Reply by PixelTim** [reply]
Yes, if you will open the .tscn file in notepad (or any other text editor program), you'll see that the nodes have only lines with parameters that are not default. Here's an example from one of my scenes:
```
[node name="RegenerationTimer" type="Timer" parent="."]
wait_time = 2.0

[node name="StunTimer" type="Timer" parent="."]
one_shot = true
```
RegenerationTimer don't have `one_shot = ...` line because it uses the default value (false). StunTimer don't have `wait_time = ...` line because it uses default value (1.0)

**Reply by AThousandShips** [reply]
That assumes that the converted project has *only* old files, and would have to be run manually *before* opening *any* files in the editor, or it would not guarantee that the assumption is correct
